File: September-15-2025/sortByFreq.js
// 451. Sort Characters By Frequency
// Solved
// Medium
// Topics
// premium lock icon
// Companies
// Given a string s, sort it in decreasing order based on the frequency of the characters. The frequency of a character is the number of times it appears in the string.

// Return the sorted string. If there are multiple answers, return any of them.

// Example 1:

// Input: s = "tree"
// Output: "eert"
// Explanation: 'e' appears twice while 'r' and 't' both appear once.
// So 'e' must appear before both 'r' and 't'. Therefore "eetr" is also a valid answer.

var s = "tree"

const sortByFreq = (s) => { 
  var freq = new Map()

  for(let char of s){
    freq.set(char, (freq.get(char) || 0) + 1)
  }

  let arr = [...freq.entries()].sort((a, b) => b[1] - a[1])
  
  let result = ''

  for(let [char, count] of arr) { 
    result += char.repeat(count)
  }
  return result
}

console.log(sortByFreq(s))


// okay i have a key value pair of the letter and frequencies 

File: 2025/November-13-2025/sortByFreq.py
# 451. Sort Characters By Frequency
# Solved
# Medium
# Topics
# premium lock icon
# Companies
# Given a string s, sort it in decreasing order based on the frequency of the characters. The frequency of a character is the number of times it appears in the string.

# Return the sorted string. If there are multiple answers, return any of them.

 

# Example 1:

s = "tree"

def sortByFreq(s: str) -> str: 
  count = {}
  freq = [[] for i in range(0, len(s) + 1)]

  for letter in s: 
    count[letter] = count.get(letter, 0) + 1
  
  for letter in count:
    freq[count[letter]].append(letter)

  res = []

  for f in range(len(s), 0, -1):
    for letter in freq[f]:
      res.append(f * letter)
  return ''.join(res)


print(sortByFreq(s))